### ISSUE TITLE
Proposal: @HasRole and @HasPermission annotations

### ISSUE THREAD

**Issue by wardr** [pac4j/jax-rs-pac4j#11]
Currently,  resources are protected via authorizers that are created on application startup, e.g. `@Pac4JSecurity(authorizers = "foo")`.

A much more convenient way of doing this would be to have two annotations:

```
public @interface HasRole {
    String value();
}

public @interface HasPermission {
    String value();
}
```

So you could have annotations like @HasRole("superuser") or @HasPermission("admin:view").

**Comment by victornoel** [comment]
I will find some time to implement it soon :)

**Comment by wardr** [comment]
Just wanted to add, now that I better understand pac4j, I think this improvement would be a big step forward towards making the developer experience with pac4j better, especially if it can use the default client to make it as simple as possible to use.

The reasons I believe these annotations are a big improvement are:

1) Only need to declare the role/permission you want to filter on in one place. 

```
// Set up authorizer in bundle
config.addAuthorizer("canEditUser", new RequireAnyPermissionAuthorizer<>("user:edit"));

// Then, in the resource class -- and this might be the only place we use this...
@Pac4JSecurity(authorizer = "canEditUser" client="form")
```

-- vs -- 

`@HasPermission("user:edit") `

2) Clarity/less magic. A newcomer needs to know that they have to set up the authorizers in the bundle, and that they need to specify which client to use, and they need to remember to use the name they associated with it in the confi, etc. In short: they need to know more about the internal plumbing of pac4j just to get started. 

3) makes it easier to use fine-grained permissions

**Comment by leleuj** [comment]
The pac4j design is merely driven by the fact it must work for 14 implementations and more. Hence its generic nature and the easy switch between one framework to another.

I understand your point of view, but I'm not so enthusiastic about this syntactic sugar. For me, it's another thing to learn: what happens for example if I encounter this annotation without being authenticated? Does it have the same behavior than the `Pac4jSecurity` annotation?

Overall, I think there is a balance to find even it seems reasonable to think about this trivial annotations.

**Comment by victornoel** [comment]
In my head, these annotations would have been used in conjunction with a `Pac4jSecurity` since you still need to specify the clients to be used.
Or maybe they should be a replacement (and thus the clients must be specified next to the authorization rule).

It's true that it can make things a bit more complex to understand (even though simpler to express).

I don't really know what is best, on one hand, it's not really needed to do things this way since pac4j is based on a global configuration approach; on the other hand, it is common to have this kind of annotations for jax-rs related security frameworks and we already started going this way with the `Pac4jSecurity` and co annotations...

**Comment by wardr** [comment]
My suggestions come from my view that for a young project like pac4j, anything that helps people have a good experience with it out of the box without having to invest too much first is important to increasing adoption.  A big part of that is making it really, really easy to get started, and gradually engage developers with a richer set of tools as more complex requirements arise. 

Both of you are far, far more familiar with this domain than I, so I certainly won't presume that my suggestions are the right way to do this, I just want to encourage you to find ways to help people quickly and painlessly add security to their apps via pac4j.

**Comment by leleuj** [comment]
Sure, it makes sense to have these new annotations. I just wanted to highlight the fact that these additions must be wisely thought.

**Comment by victornoel** [comment]
I won't have the time to tackle that soon, so let's postpone it and release a new version soon before starting to move everything to pac4j v2.

**Comment by leleuj** [comment]
Closing due to inactivity.
